File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        HashMap<String, Integer> map = new HashMap<>();
        
        for(int i = 0; i < players.length; i++) map.put(players[i], i);
        for(String calling : callings) {
            int idx = map.get(calling);
            String left = players[idx - 1];
            String right = players[idx];
            map.put(left, idx);
            map.put(right, idx - 1);
            players[idx - 1] = right;
            players[idx] = left;
        }
        
        return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    const map = new Map();
    
    for (i = 0; i < players.length; i++) {
        const player = players[i];
        map.set(player, i);
    }
    for (calling of callings) {
        const playerIdx = map.get(calling);
        const targetPlayerIdx = playerIdx - 1;
        map.set(calling, targetPlayerIdx);
        map.set(players[targetPlayerIdx], playerIdx);
        swapRanking(players, playerIdx, targetPlayerIdx);
    }
    
    return players;
}

const swapRanking = (players, playerIdx, targetPlayerIdx) => {
    const tmp = players[playerIdx];
    players[playerIdx] = players[targetPlayerIdx];
    players[targetPlayerIdx] = tmp;
}